File: src/main/java/spring/springserver/global/jwt/MemberDetails.java
package spring.springserver.global.jwt;

import jakarta.validation.constraints.Email;
import org.jspecify.annotations.NonNull;
import org.jspecify.annotations.Nullable;
import org.springframework.security.core.GrantedAuthority;
import org.springframework.security.core.authority.SimpleGrantedAuthority;
import org.springframework.security.core.userdetails.UserDetails;
import spring.springserver.domain.member.entity.Member;

import java.util.Collection;
import java.util.List;

public record MemberDetails(Long id,
							String username,
							@Email String email,
							String password,
							Collection<? extends GrantedAuthority> authorities) implements UserDetails {

	public static MemberDetails from(Member member) {
		String role = member.getRole().name();
		if (!role.startsWith("ROLE_")) {
			role = "ROLE_" + role;
		}

		return new MemberDetails(
				member.getId(),
				member.getUsername(),
				member.getEmail(),
				member.getPassword(),
				List.of(new SimpleGrantedAuthority(role))
		);
	}

	@Override
	public @NonNull Collection<? extends GrantedAuthority> getAuthorities() {
		return authorities;
	}

	@Override
	public @Nullable String getPassword() {
		return password();
	}

	@Override
	public @NonNull String getUsername() {
		return username();
	}
}

File: src/main/java/spring/springserver/global/jwt/MemberDetails.kt
package spring.springserver.global.jwt

import jakarta.validation.constraints.Email
import org.springframework.security.core.GrantedAuthority
import org.springframework.security.core.authority.SimpleGrantedAuthority
import org.springframework.security.core.userdetails.UserDetails
import spring.springserver.domain.member.entity.Member

data class MemberDetails(private val id: Long?,
                         private val username: String,
                         @field:Email private val email: String,
                         private val password: String,
                         private val authorities: Collection<GrantedAuthority>): UserDetails {

    override fun getAuthorities(): Collection<GrantedAuthority?> = authorities

    override fun getPassword(): String = password

    override fun getUsername(): String = username

    companion object {

        fun from(member: Member): MemberDetails {

            var role = member.role.name

            if (!role.startsWith("ROLE_")) {

                role = "ROLE_$role"
            }

            return MemberDetails(
                member.getId(),
                member.username,
                member.email,
                member.password,
                listOf(SimpleGrantedAuthority(role))
            )
        }
    }
}
